ApplyBQSRDataflow, with passing tests.

jean-philippe-martin opened the pull request:
This includes the workaround for https://github.com/google/google-http-java-client/issues/293

jean-philippe-martin commented:
It took:
9m30s for the 10 tests in BaseRecalibratorDataflowIntegrationTest, and
6m47s for the 2 tests in ApplyBQSRDataflowIntegrationTest.

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/ApplyBQSRTransform.java:
```diff
@@ -42,19 +41,24 @@ public ApplyBQSRTransform(SAMFileHeader header, PCollection<BaseRecalOutput> rec
         @Override
         public void processElement(ProcessContext c) {
             if (null==transformer) {
                 // set up the transformer, as needed
```
Remove uninformative comment

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/ApplyBQSRTransform.java:
```diff
@@ -1,14 +1,13 @@
 package org.broadinstitute.hellbender.dev.pipelines.bqsr;
```
Move out of dev and into tools.dataflow.transforms.bqsr

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/tools/walkers/bqsr/ApplyBQSRDataflow.java:
```diff
@@ -1,7 +1,10 @@
 package org.broadinstitute.hellbender.dev.tools.walkers.bqsr;
```
Move out of dev and into tools.dataflow.pipelines

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/tools/walkers/bqsr/ApplyBQSRDataflow.java:
```diff
@@ -74,15 +77,16 @@ protected void setupPipeline(Pipeline pipeline) {
-        PCollection<GATKRead> output = readsSource.getReadPCollection(intervals, ValidationStringency.SILENT, false)
-                .apply(new ApplyBQSRTransform(header, recalInfoSingletonCollection, bqsrOpts));
+        final PCollectionView<BaseRecalOutput> recalInfoSingletonView = BaseRecalOutputSource.loadFileOrRemote(pipeline, BQSR_RECAL_FILE_NAME).apply(View.asSingleton());
+                PCollection<GATKRead> output = readsSource.getReadPCollection(intervals, ValidationStringency.SILENT, false)
```
Fix indentation here

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/ReadsPreprocessingPipeline.java:
```diff
@@ -115,10 +117,10 @@ protected void setupPipeline( Pipeline pipeline ) {
-
         final PCollectionView<BaseRecalOutput> mergedRecalibrationReport = recalibrationReports.apply(View.<BaseRecalOutput>asSingleton());
 
-        final PCollection<GATKRead> finalReads = markedReads.apply(new ApplyBQSRStub(headerSingleton, mergedRecalibrationReport));
```
Delete the `ApplyBQSRStub` as well

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/read/GoogleGenomicsReadToGATKReadAdapter.java:
```diff
@@ -631,6 +633,16 @@ public void clearAttributes() {
+        }
+        genomicsRead.setInfo(infoCopy);
+        genomicsRead.getAlignment().setCigar(new ArrayList<>(genomicsRead.getAlignment().getCigar()));
+        // end workaround
```
This workaround shouldn't be necessary given https://github.com/broadinstitute/hellbender/pull/801. Could you please delete it and confirm that tests still pass?

jean-philippe-martin replied:
They don't. I tried that in an earlier build [7a2a831](https://github.com/broadinstitute/hellbender/commit/7a2a8319bcec80d19e6a1aeb2cf61f6f831fddbe). Here's the [Travis report](https://travis-ci.org/broadinstitute/hellbender/jobs/76221794). It cuts off the good parts and I didn't keep my own copy, but there was some other type that the reflection-based serialization had trouble with.

droazen replied:
In that case, could you please:

-Move the workaround into its own private method `workaroundHttpClientCloneBug()` called from `copy()`, with a comment to delete once the fix is in.

-Alter the workaround so that it won't fail if `genomicsRead.getInfo() == null` or `genomicsRead.getAlignment() == null` or `genomicsRead.getAlignedQuality() == null`

droazen replied:
Also, could you please file a bug with google-http-client-java explaining that this is still a problem (since we are using the snapshot with the previous fix as of https://github.com/broadinstitute/hellbender/pull/801), and request a **general** solution to the problem of instantiating private types (ie., use reflection to `getDeclaredConstructor()` and `constructor.setAccessible(true)`)

droazen commented:
Review complete -- back to @jean-philippe-martin. Feel free to rebase/merge after addressing comments.

jean-philippe-martin commented:
Applied feedback, reproduced bug and updated our description (#650), submitted bug report (https://github.com/google/google-http-java-client/issues/297), squashed. Merging once tests pass.
